### PR TITLE
AP_Common: Location.cpp: add sanity checks

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

**Comment by peterbarker** [comment]
These sanity checks are catching things in CI.

**Comment by khancyr** [comment]
return false as it is done currently ;-P

File: libraries/AP_Common/Location.cpp [hunk 1]
```diff
@@ -158,7 +163,9 @@ bool Location_Class::get_alt_cm(ALT_FRAME desired_frame, int32_t &ret_alt_cm) co
             alt_abs = alt + alt_terr_cm;
             break;
         default:
-            // unknown conversion to absolute alt, this should never happen
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+            AP_HAL::panic("unknown conversion to absolute alt, this should never happen");
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/AP_Common/Location.cpp]: Done

**Comment by OXINARF** [comment]
> These sanity checks are catching things in CI.

It's catching one of those, just don't know where 😛

**Comment by peterbarker** [comment]
@OXINARF After the slog to find/fix the first one I was somewhat dismayed to get hit by another one :-)

**Comment by peterbarker** [comment]
OK, so this one was easier to find.

A test is setting the switch position to RTL.

We then reboot the vehicle.

Because the motors are disarmed at this point, we are ignore any and all checks when changing mode.

That forces us into RTL on reboot.  We then go and build a return path to 0,0 as that's the home location....
but this new sanity check whinges about working with a crap location....

I could add a check in so we don't build that home path or whatever - I just have a feeling this is going to be tip-of-the-iceberg with the whole "ignore checks when disarmed" thing.

**Comment by peterbarker** [comment]
I've pushed up a patch which stops the check whinging when forced into RTL mode.

File: ArduCopter/mode_rtl.cpp [hunk 2]
```diff
@@ -65,6 +75,11 @@ void Copter::ModeRTL::run(bool disarm_on_land)
     // call the correct run function
     switch (_state) {
 
+    case RTL_Starting:
+        // should not be reached:
+        _state = RTL_InitialClimb;
+        // FALLTHROUGH
```
Comment (WickedShell) [inline ArduCopter/mode_rtl.cpp]: Wrong fallthrough, should be a non comment `FALLTHROUGH;`

Comment (peterbarker) [reply ArduCopter/mode_rtl.cpp]: Fixed.

**Comment by peterbarker** [comment]
@khancyr good to go - apart from the fact that it still panics because the sanity check is doing its job :-)

**Comment by magicrub** [comment]
looks OK to me now, ok to merge?

**Comment by peterbarker** [comment]
On Thu, 14 Mar 2019, Tom Pittenger wrote:

> looks OK to me now, ok to merge?

I think it's OK apart from the panic() when someone directs SITL to 
navigate to 0,0,0 - like the mount test does...

I've pushed up a patch for that...

**Comment by rmackay9** [comment]
Unfortunately this also needs to wait on spool logic going in (sorry!): https://github.com/ArduPilot/ardupilot/pull/10656

**Comment by rmackay9** [comment]
as discussed on the dev call we should rebase this and I thought that our position_ok() required a home to be set (but perhaps I'm wrong).

**Comment by peterbarker** [comment]
@rmackay9 `position_ok` doesn't require home be set, 'though its comment in the header seems to indicate that was expected.  I went back as far as 3.3, and I didn't see the requirement in code...

This branch: https://github.com/peterbarker/ardupilot/tree/wip/location-private-alt makes the alt field private, and replaces `current_loc.alt` with a method to call if you believe home should be set.  Currently I'm looking at this stack trace:

```
(gdb) bt
#0  AP_HAL::panic (errormsg=0x555555738120 "internal error %u")
    at ../../libraries/AP_HAL_SITL/system.cpp:34
#1  0x00005555555fe973 in AP_InternalError::error (
    this=0x5555559a9c08 <instance>, 
    e=AP_InternalError::error_t::home_relalt_no_home)
    at ../../libraries/AP_InternalError/AP_InternalError.cpp:8
#2  0x00005555555a4b55 in Copter::CurrentLocation::safe_home_relative_alt (
    this=0x5555559a76d4 <copter+15956>) at ../../ArduCopter/navigation.cpp:41
#3  0x000055555559123d in Copter::ModeAuto::takeoff_start (
    this=0x5555559a8a28 <copter+20904>, dest_loc=...)
    at ../../ArduCopter/mode_auto.cpp:160
#4  0x0000555555593083 in Copter::ModeAuto::do_takeoff (
    this=0x5555559a8a28 <copter+20904>, cmd=...)
    at ../../ArduCopter/mode_auto.cpp:1071
#5  0x0000555555591abf in Copter::ModeAuto::start_command (
    this=0x5555559a8a28 <copter+20904>, cmd=...)
    at ../../ArduCopter/mode_auto.cpp:410
#6  0x000055555558058d in Functor<bool, AP_Mission::Mission_Command const&>::method_wrapper<Copter::ModeAuto, &Copter::ModeAuto::start_command> (
    obj=0x5555559a8a28 <copter+20904>, args#0=...)
    at ../../libraries/AP_HAL/utility/functor.h:88
#7  0x00005555555e50a1 in Functor<bool, AP_Mission::Mission_Command const&>::operator() (this=0x5555559a8ab8 <copter+21048>, args#0=...)
---Type <return> to continue, or q <return> to quit---
    at ../../libraries/AP_HAL/utility/functor.h:54
#8  0x00005555555e1883 in AP_Mission::start_command (
    this=0x5555559a8aa8 <copter+21032>, cmd=...)
    at ../../libraries/AP_Mission/AP_Mission.cpp:298
#9  0x00005555555e4449 in AP_Mission::advance_current_nav_cmd (
    this=0x5555559a8aa8 <copter+21032>, starting_index=0)
    at ../../libraries/AP_Mission/AP_Mission.cpp:1511
#10 0x00005555555e1205 in AP_Mission::start (
    this=0x5555559a8aa8 <copter+21032>)
    at ../../libraries/AP_Mission/AP_Mission.cpp:72
#11 0x00005555555e1288 in AP_Mission::resume (
    this=0x5555559a8aa8 <copter+21032>)
    at ../../libraries/AP_Mission/AP_Mission.cpp:100
#12 0x00005555555e1478 in AP_Mission::start_or_resume (
    this=0x5555559a8aa8 <copter+21032>)
    at ../../libraries/AP_Mission/AP_Mission.cpp:168
#13 0x0000555555590fc5 in Copter::ModeAuto::init (
    this=0x5555559a8a28 <copter+20904>, ignore_checks=true)
    at ../../ArduCopter/mode_auto.cpp:47
#14 0x000055555558de85 in Copter::set_mode (this=0x5555559a3880 <copter>, 
    mode=AUTO, reason=MODE_REASON_TX_COMMAND) at ../../ArduCopter/mode.cpp:227
#15 0x0000555555586a08 in RC_Channel_Copter::mode_switch_changed (
    this=0x5555559a3c28 <copter+936>, new_pos=3 '\003')
---Type <return> to continue, or q <return> to quit---
    at ../../ArduCopter/RC_Channel.cpp:23
#16 0x0000555555629df0 in RC_Channel::read_mode_switch (
    this=0x5555559a3c28 <copter+936>)
    at ../../libraries/RC_Channel/RC_Channel.cpp:412
#17 0x000055555562ae29 in RC_Channels::read_mode_switch (
    this=0x5555559a3b50 <copter+720>)
    at ../../libraries/RC_Channel/RC_Channels.cpp:195
#18 0x000055555557b8fe in Copter::rc_loop (this=0x5555559a3880 <copter>)
    at ../../ArduCopter/ArduCopter.cpp:276
#19 0x000055555557c558 in Functor<void>::method_wrapper<Copter, &Copter::rc_loop> (obj=0x5555559a3880 <copter>) at ../../libraries/AP_HAL/utility/functor.h:88
#20 0x00005555555e50c4 in Functor<void>::operator() (
    this=0x555555998de0 <Copter::scheduler_tasks>)
    at ../../libraries/AP_HAL/utility/functor.h:54
#21 0x00005555555f9b2c in AP_Scheduler::run (
    this=0x5555559a4178 <copter+2296>, time_available=2500)
    at ../../libraries/AP_Scheduler/AP_Scheduler.cpp:170
#22 0x00005555555f9f08 in AP_Scheduler::loop (
    this=0x5555559a4178 <copter+2296>)
    at ../../libraries/AP_Scheduler/AP_Scheduler.cpp:264
#23 0x000055555557b7e0 in Copter::loop (this=0x5555559a3880 <copter>)
    at ../../ArduCopter/ArduCopter.cpp:217
#24 0x0000555555685697 in HAL_SITL::run (
---Type <return> to continue, or q <return> to quit---
    this=0x5555559b1ec0 <AP_HAL::get_HAL()::hal>, argc=9, argv=0x7fffffffda78, 
    callbacks=0x5555559a3880 <copter>)
    at ../../libraries/AP_HAL_SITL/HAL_SITL_Class.cpp:95
#25 0x000055555557c523 in main (argc=9, argv=0x7fffffffda78)
    at ../../ArduCopter/ArduCopter.cpp:591
(gdb) 
(gdb) print copter.arming.armed
$2 = false
(gdb) 
```

If one were to leave auto, wait for armability and get home set +30m, arm then return to auto and takeoff we might takeoff with bad numbers locked in.

**Comment by khancyr** [comment]
@peterbarker , was is missing to merge this ?

**Comment by peterbarker** [comment]
@khancyr waiting for @rmackay9  to respond to the position-requires-home thing.  The stack trace I pasted in is a separate issue.

Note the changes in inertia.cpp also fix an issue.

The panic in here could notionally be replaced with an internalerror - but I've got a sneaking suspicion we'll be using uninitialised locations (mostly in a benign manner) all over the place at the moment and we don't want to over-use internal-error or people will get used to it!

**Comment by rmackay9** [comment]
Merged, thanks!

**Comment by rmackay9** [comment]
So despite all the reviews by no less than 6 people, this introduced this bug into Copter.  https://github.com/ArduPilot/ardupilot/issues/13441

**Pull request merged**